inventory_plugins: Add KubeVirt inventory plugin

machacekondra opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR add inventory plugin for [kubevirt](http://kubevirt.io/).

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
kubevirt

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
2.8
```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

machacekondra committed:
inventory_plugins: Add kubevirt inventory plugin

Signed-off-by: Ondra Machacek <omachace@redhat.com>

machacekondra commented:
@pkliczewski @mmazur

pkliczewski commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,178 @@
+
+    description:
+      - Fetch VirtualMachines and VirtualMachinesInstances for one or more clusters
+      - Groups by cluster name, namespace, namespace_vms, namespace_vmis  and labels
```
What are `namespace_vms, namespace_vmis`?

machacekondra replied:
Right, `namespace_vms` should be removed .. actually even `namespace_vmis`, namespace is enough, if we don't support VMs, but only VMIs.

mmazur replied:
This still needs addressing.

mmazur replied:
(As in – we should drop the vm/vmis)

pkliczewski commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,178 @@
+        description: token that ensures this is a source file for the 'kubevirt' plugin.
+        required: True
+        choices: ['kubevirt']
+      connections:
```
I see that there is one more param in lookup params [here](https://docs.ansible.com/ansible/2.5/plugins/lookup/k8s.html) which is missing from the list. It is `kind` parameter. Should we add it here?

machacekondra replied:
That's lookup plugin not inventory plugin. It's not relevant here.

pkliczewski commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,178 @@
+            for namespace in namespaces:
+                self.get_vms_for_namespace(client, name, namespace)
+
+    def get_vms_for_namespace(self, client, name, namespace):
```
We use `vms` name whereas we fetch `vmis`. I suggest to be consistent. In addition why we are fetching vmis and not vms?

machacekondra replied:
Because VMIs has IPs, VMs don't so it doesn't make much sense to fetch them, but if you think we should, we can add it.

mmazur commented:
Please mark this as WIP, since this is nowhere near ready.

machacekondra commented:
It's ready for review. Code is working fine.

mmazur commented:
Ok, my comments to the code:
1. Since kubevirt is in development it's probably a good idea to add _api_version_ as a parameter to the module, so this keeps being usable when upstream kubevirt changes api versions.
2. The plugin description and the code state explicitly and imply that this code deals with VirtualMachines. It doesn't. It only operates on VMInstances. This should be corrected.
3. While it's ok for the initial version to only support VMIs, considering kubevirt/k8s modules allow operating on a cluster's resources directly, we might want to design it in such a way that adding listing of non–VMI objects (say: VMs) in the future does not become a problem.
4. The (namespace, resource type, name) combo is guaranteed unique. I'd suggest using `$namespace___VMI___$name` for host names instead of the IPs. This also solves previous point. (Putting `___VMI___` in the middle makes it virtually impossible for someone to accidentally cause naming conflicts with unusual namespace/vmi names.)
5. All the relevant metadata for any given VMI should be readily available. For now stuff like IP as `ansible_host` and object type as `object_type` (== `VMI`) are likely sufficient. If we ever add say VM support, then `owner` (e.g. == `default___VM___test-fedora`) and others would likely be of use.

pkliczewski commented:
> Ok, my comments to the code:
> 
> 1. Since kubevirt is in development it's probably a good idea to add _api_version_ as a parameter to the module, so this keeps being usable when upstream kubevirt changes api versions.

This needs to be verified. With any api (breaking?) changes we need to verify. I would rather make the code fail due to incompatible version than fail later during invocation of the api.

> 2. The plugin description and the code state explicitly and imply that this code deals with VirtualMachines. It doesn't. It only operates on VMInstances. This should be corrected.

I would hide this implementation detail from users. What they want is `VirtualMachines`.

> 3. While it's ok for the initial version to only support VMIs, considering kubevirt/k8s modules allow operating on a cluster's resources directly, we might want to design it in such a way that adding listing of non–VMI objects (say: VMs) in the future does not become a problem.

Agree

> 4. The (namespace, resource type, name) combo is guaranteed unique. I'd suggest using `$namespace___VMI___$name` for host names instead of the IPs. This also solves previous point. (Putting `___VMI___` in the middle makes it virtually impossible for someone to accidentally cause naming conflicts with unusual namespace/vmi names.)

In general `namespace` and `name` should be unique for specific object `kind`. We may check whether  an object exists and fail. I would not add any other means to improve uniqueness done in k8s. 

> 5. All the relevant metadata for any given VMI should be readily available. For now stuff like IP as `ansible_host` and object type as `object_type` (== `VMI`) are likely sufficient. If we ever add say VM support, then `owner` (e.g. == `default___VM___test-fedora`) and others would likely be of use.

As in point 4.

mmazur commented:
> This needs to be verified. With any api (breaking?) changes we need to verify. I would rather make the code fail due to incompatible version than fail later during invocation of the api.

If the api version change breaks compatibility, then trying to run this module with a non–default api_version will give you an error (most likely some kind of "attribute XY not found").
If the api version does not break compatibility, then this inventory module keeps being usable with different kubevirt releases.

The way I see it is there's a major upside with no downsides, since if it does fail, it'll do so explicitly and quite loudly.
 
> > 1. The plugin description and the code state explicitly and imply that this code deals with VirtualMachines. It doesn't. It only operates on VMInstances. This should be corrected.
> 
> I would hide this implementation detail from users. What they want is `VirtualMachines`.

Yes, but this module as it exists does not operate on VirtualMachines, but VirtualMachineInstances. If you have a bunch of vm instances running, some ephemeral, some not and some parts of replicasets, this module will return them all and claim they're "vms".

This is fine if we assume upfront that the only thing this module will ever do is return a list of hosts you can tcp connect to. But if we ever wanted to use this to get a list of VMReplicaSets or only non–ephemeral vms, etc., then we should make it way more explicit that the current code only deals with VMIs, or we're gonna have a bad time extending it further down the line.

> > 1. The (namespace, resource type, name) combo is guaranteed unique. I'd suggest using `$namespace___VMI___$name` for host names instead of the IPs. This also solves previous point. (Putting `___VMI___` in the middle makes it virtually impossible for someone to accidentally cause naming conflicts with unusual namespace/vmi names.)
> 
> In general `namespace` and `name` should be unique for specific object `kind`. We may check whether an object exists and fail. I would not add any other means to improve uniqueness done in k8s.

My suggestion was for what to call "hosts" returned by this plugin. Currently it just uses their IPs as hostnames, which means it can only return a list of currently running VMIs. Again, this is fine if we want to only be able to list stuff one can connect to and never be able to operate on kubevirt's objects. (Dynamically add/remove labels, change number of replicas, etc.)

pkliczewski commented:
> If the api version change breaks compatibility, then trying to run this module with a non–default api_version will give you an error (most likely some kind of "attribute XY not found").

This was the case so far. The version was changed due to breaking changes. I think that we need to be explicit with which version we are working and do not let the user use the module with not correct version.

> This is fine if we assume upfront that the only thing this module will ever do is return a list of hosts you can tcp connect to. But if we ever wanted to use this to get a list of VMReplicaSets or only non–ephemeral vms, etc., then we should make it way more explicit that the current code only deals with VMIs, or we're gonna have a bad time extending it further down the line.

So far we defined it as:

> you may use it to dynamically manage vms

I think we need to be more specific about what we want to achieve by this module.

> My suggestion was for what to call "hosts" returned by this plugin. Currently it just uses their IPs as hostnames, which means it can only return a list of currently running VMIs. Again, this is fine if we want to only be able to list stuff one can connect to and never be able to operate on kubevirt's objects. (Dynamically add/remove labels, change number of replicas, etc.)

As above we need to be more specific about the scope. 

@machacekondra please update Trello card with it. It is fine to work with `vmi`s for now.

machacekondra commented:
I've just found out there are many ways of exposing IPs as services for different project. So may be worth to rethink this plugin.

mmazur commented:
I'm perfectly fine with supporting only VMIs in this PR, but afaik it's considered bad practice to use IPs for ansible inventory host names (especially when we have unique identifiers). And will make it harder to support anything else. So I'd like to see at least that modified.

machacekondra commented:
```bash
omachace ~/ansible/kubevirt $ cat kubevirt.yaml 
plugin: kubevirt
connections:
  - namespaces:
      - default
    network_name: default
omachace ~/ansible/kubevirt $ ansible-inventory -i kubevirt.yaml --list
{
    "10-37-140-246_6443": {
        "children": [
            "namespace_default"
        ]
    },
    "_meta": {
        "hostvars": {
            "10.244.0.10": {
                "annotations": {},
                "labels": {
                    "kubevirt.io/nodeName": "omachace-master",
                    "special": "testvm"
                },
                "object_type": "vm",
                "resource_version": "436650",
                "uid": "0b16d34e-01eb-11e9-b2c0-001a4a013f37"
            }
        }
    },
    "all": {
        "children": [
            "10-37-140-246_6443",
            "label_kubevirt.io/nodeName_omachace-master",
            "label_special_testvm",
            "ungrouped"
        ]
    },
    "label_kubevirt.io/nodeName_omachace-master": {
        "hosts": [
            "10.244.0.10"
        ]
    },
    "label_special_testvm": {
        "hosts": [
            "10.244.0.10"
        ]
    },
    "namespace_default": {
        "children": [
            "namespace_default_vms"
        ]
    },
    "namespace_default_vms": {
        "hosts": [
            "10.244.0.10"
        ]
    },
    "ungrouped": {}
}
```

mmazur commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,198 @@
+    short_description: KubeVirt inventory source
+
+    description:
+      - Fetch VirtualMachines and VirtualMachinesInstances for one or more clusters
```
Description should be: _Fetch running virtual machines from KubeVirt clusters_

Reasoning:
1. This only returns *running* vms.
2. We've decided to hide the internal implementation of VMs/VMIs (as much as possible) from ansible users.

mmazur commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,198 @@
+              description:
+              - List of namespaces. If not specified, will fetch all virtual machines for all namespaces user is authorized
+                to access.
+          network_name:
```
I don't know what the opinion of the upstream reviewers will be, but considering this parameter controls a host's inventory hostname and not just `ansible_host`, it might be the case that this should be somehow contained within the standard inventory plugin `inventory_host` parameter.

machacekondra replied:
This parameter filters the network which should be used as a primary IP address.

machacekondra commented:
Does it even make sense to merge in 2.8? Should we start following this guidlines: https://github.com/ansible-collections/ ?

pkliczewski commented:
it is our goal to merge it in 2.8. Let's talk about it whether we will have enough time to use `ansible-collections`.

machacekondra commented:
@mattclay @gundalow Can you please review and if possible merge into 2.8?

pkliczewski commented:
@mattclay @gundalow please review

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,201 @@
+        description: token that ensures this is a source file for the 'kubevirt' plugin.
+        required: True
+        choices: ['kubevirt']
+      connections:
```
missing 'type: list'

bcoca replied:
actually most options are missing 'type'  and none show if there is a default or is not required 
are they all required always?

machacekondra replied:
This is actually list of dictionaries. So I can really ell document the values inside the dictionary, because it's not visible in generated documentation.

bcoca replied:
ansible-doc shows it correctly, its also not just for 'documentation' it also controls how to pass parameters to the plugin itself and validation of those parameters

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,201 @@
+              - In case of multiple network attached to virtual machine, define which interface should be returned as primary IP
+                address.
+    requirements:
+    - "python >= 2.7"
```
this is already a requirement for the Ansible controller, you can omit it

machacekondra commented:
@bcoca Can you please re-review?

machacekondra commented:
ping

machacekondra commented:
@bcoca Can you please re-check?

machacekondra commented:
@bcoca Can you please review - so the plugin can go to 2.8 release please?

thedoubl3j reviewed (commented):
first pass is that this looks like it is in a good state. the idea of supporting specific API versions I think is the correct approach and thanks for adding in `type:`. 

lgtm

edit: grammar is hard

jctanner commented:
bot_status

pkliczewski commented:
shipit

goneri commented:
:+1:

goneri commented:
I just deployed kubevirt ( http://kubevirt.io/get_kubevirt/ ) on a minikube:

```shell
$ kubectl get vms
NAME        AGE   RUNNING   VOLUME
vm-cirros   20m   true
$ ansible-inventory -i ~/kubevirt.yaml --list
{
    "192-168-39-118_8443": {
        "children": [
            "namespace_default"
        ]
    },
    "_meta": {
        "hostvars": {}
    },
    "all": {
        "children": [
            "192-168-39-118_8443",
            "ungrouped"
        ]
    },
    "namespace_default": {
        "children": [
            "namespace_default_vms"
        ]
    }
}
$ kubectl get namespaces
NAME          STATUS   AGE
default       Active   43m
kube-public   Active   43m
kube-system   Active   43m
kubevirt      Active   40m
$ cat ~/kubevirt.yaml
plugin: kubevirt
connections:
  - namespaces:
      - default
    network_name: default
```

I tried different configuration with my `~/kubevirt.yaml`, but I always get this empty output. Any idea why  I did wrong?

goneri commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,214 @@
+class InventoryModule(K8sInventoryModule):
+    NAME = 'kubevirt'
+
+    transport = 'oc'
```
This variable is never used in the code.

machacekondra replied:
Right, will remove.

machacekondra commented:
Seems that your VM does not have any network interface, which report IP address, can you please post output of the following command:

```
oc get vmis -o yaml
```

mmazur commented:
@machacekondra it'd be a good idea if you used something other than ips for hostnames.

goneri commented:
@machacekondra indeed, I used the cluster/examples/vm-cirros.yaml template. And I do see any network configuration. I was naively expecting the VM to be reachable using the oc ( https://docs.ansible.com/ansible/latest/plugins/connection/oc.html ) transport.

goneri commented:
Ok, this time I managed to everything working just fine. I agree with @mmazur, it would be much better to put the IP address in `ansible_host` key and use the VM name as the Ansible host name.

mmazur commented:
Not gonna dive into the code right now to check, but dependent on how the plugin deals with namespaces, using namespace+name might be the way to go.

goneri commented:
I understand the idea, but I think this kind of behaviour should also be configurable and maybe disabled by default.

mmazur commented:
Other approach: the vmware inv plugin uses name+uuid to guarantee uniqueness.

machacekondra commented:
Ok, added grouping by VM name, and added variable `ansible_host`. @mmazur hostname isn't reliable IMHO for this, because it's set only when cloud-init is used AFAIK, I could create some fallback mechanism to try hostname in case of cloud-init is used, but not now imho.

machacekondra commented:
We create groups based on namespaces and labels.

machacekondra commented:
The current inv looks like this(two namespaces and some labels):

```json
{
    "192-168-99-100_8443": {
        "children": [
            "namespace_default",
            "namespace_myproject"
        ]
    },
    "_meta": {
        "hostvars": {
            "default.myvm": {
                "annotations": {},
                "ansible_host": "172.17.0.20",
                "labels": {
                    "app": "postgres",
                    "kubevirt.io/nodeName": "localhost"
                },
                "object_type": "vm",
                "resource_version": "8040",
                "uid": "c8a61264-464e-11e9-bace-0800279ffc6b"
            },
            "default.myvm2": {
                "annotations": {},
                "ansible_host": "172.17.0.19",
                "labels": {
                    "app": "http",
                    "kubevirt.io/nodeName": "localhost"
                },
                "object_type": "vm",
                "resource_version": "8304",
                "uid": "e2ee9ab7-464e-11e9-bace-0800279ffc6b"
            },
            "myproject.myvm": {
                "annotations": {},
                "ansible_host": "172.17.0.22",
                "labels": {
                    "app": "postgres",
                    "kubevirt.io/nodeName": "localhost"
                },
                "object_type": "vm",
                "resource_version": "8653",
                "uid": "0197e085-464f-11e9-bace-0800279ffc6b"
            },
            "myproject.myvm3": {
                "annotations": {},
                "ansible_host": "172.17.0.21",
                "labels": {
                    "app": "postgres",
                    "kubevirt.io/nodeName": "localhost"
                },
                "object_type": "vm",
                "resource_version": "8511",
                "uid": "f679e33f-464e-11e9-bace-0800279ffc6b"
            }
        }
    },
    "all": {
        "children": [
            "192-168-99-100_8443",
            "label_app_http",
            "label_app_postgres",
            "label_kubevirt.io/nodeName_localhost",
            "ungrouped"
        ]
    },
    "label_app_http": {
        "hosts": [
            "default.myvm2"
        ]
    },
    "label_app_postgres": {
        "hosts": [
            "default.myvm",
            "myproject.myvm",
            "myproject.myvm3"
        ]
    },
    "label_kubevirt.io/nodeName_localhost": {
        "hosts": [
            "default.myvm",
            "default.myvm2",
            "myproject.myvm",
            "myproject.myvm3"
        ]
    },
    "namespace_default": {
        "children": [
            "namespace_default_vms"
        ]
    },
    "namespace_default_vms": {
        "hosts": [
            "default.myvm",
            "default.myvm2"
        ]
    },
    "namespace_myproject": {
        "children": [
            "namespace_myproject_vms"
        ]
    },
    "namespace_myproject_vms": {
        "hosts": [
            "myproject.myvm",
            "myproject.myvm3"
        ]
    }
}
```

bcoca commented:
this should not be a valid group name ` "label_kubevirt.io/nodeName_localhost"`

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,219 @@
+
+    short_description: KubeVirt inventory source
+
+    description:
```
missing `version_added`

machacekondra replied:
done

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,219 @@
+          - Optional list of cluster connection settings. If no connections are provided, the default
+            I(~/.kube/config) and active context will be used, and objects will be returned for all namespaces
+            the active user is authorized to access.
+          name:
```
you might want a `suboptions` key here,  they are not implemented in config (yet) but eventually they should work the same as module suboptions.

machacekondra replied:
done

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,219 @@
+              - In case of multiple network attached to virtual machine, define which interface should be returned as primary IP
+                address.
+              type: str
+          api_version:
```
is this per connection or global? some options seem like they are overtly indented

machacekondra replied:
per connection, we have only one top level parameter `connections`

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,219 @@
+        namespace_group = 'namespace_{0}'.format(namespace)
+        namespace_vms_group = '{0}_vms'.format(namespace_group)
+
+        self.inventory.add_group(name)
```
`add_group` returns a sanitized name, to ensure the `add_child` below doesn't error out either use that name or sanitize `name` beforehand. You can use `self._sanitize_group_name()`

machacekondra replied:
done

bcoca reviewed (changes requested):
You are also missing a `verify_file` which means you will accept any file name/extensions, we try to shy away from that since it has caused many issues with script/ini/yaml ... but we have had to keep those due to backwards compatibility.

machacekondra commented:
> this should not be a valid group name "label_kubevirt.io/nodeName_localhost"

Actually 'kubevirt.io/nodeName' is name of the label in kubernetes, I can replace all occurrences of '/' by some valid character in labels.

bcoca commented:
@machacekondra  see my comment about `self._sanitize_group_name()`

bcoca commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,219 @@
+API_VERSION = 'kubevirt.io/v1alpha3'
+
+
+class InventoryModule(K8sInventoryModule):
```
you are also missing doc fragments for classes using constructable and cacheable

machacekondra replied:
This is done as part of inheritance from `K8sInventoryModule`
`
lib/ansible/plugins/inventory/k8s.py:class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable, K8sAnsibleMixin):
`

bcoca replied:
@machacekondra still, it requires the docs

also that module would not have passed review, it got merged w/o it so it is not a good example to follow

goneri replied:
It's addressed here https://github.com/ansible/ansible/pull/47803/files#diff-4e0a09d18f6dfe19eb9d0b30c0a53bafR16

goneri commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,237 @@
+            if interface is None or interface.ipAddress is None:
+                continue
+
+            vm_name = '{0}.{1}'.format(vm.metadata.namespace, vm.metadata.name)
```
The use of the '.' as a separator sounds a bit weird to me. It sounds like "name" is a domain and namespace is hostname. I suggest to add a new parameter `name_format`, with something like that by default:

```python
name_format='{namespce}-{name}'
```

And use it here:
```python
vm_name = name_format.format(namespace=vm.metadata.namespace, name=vm.metadata.name, uid=vm.metadata.uid)
```

machacekondra replied:
Added.

goneri commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,238 @@
+            if interface is None or interface.ipAddress is None:
+                continue
+
+            name_format = '{namespace}-{name}-{uid}'
```
Could you move this string in the parameter list of the module?

machacekondra replied:
Done

goneri replied:
Thank you @machacekondra!

machacekondra commented:
@bcoca Fixed all your comments.

machacekondra commented:
ping

gundalow commented:
Rather than `ping`'ing please do `@person I've done X, could you please Y`

pkliczewski commented:
@bcoca @gundalow @goneri please review.
Shippable failure seems not to be related to this PR.

gundalow commented:
@pkliczewski I've triggered a rerun on Shippable

pkliczewski commented:
@bcoca @gundalow @goneri please review again. Shippable failure not related to this PR.

pkliczewski commented:
@bcoca @gundalow @goneri Is there anything else we need to do?

goneri commented:
I just retested the patch, everything looks good to me.

mattclay commented on test/integration/targets/inventory_kubevirt_conformance/runme.sh:
```diff
@@ -0,0 +1,58 @@
+source virtualenv.sh
+
+virtualenv --system-site-packages --python "${ANSIBLE_TEST_PYTHON_INTERPRETER:-python}" "${OUTPUT_DIR}/kubevirt-inventory"
+source "${OUTPUT_DIR}/kubevirt-inventory/bin/activate"
```
With `source virtualenv.sh` these two lines for creating and activating the virtualenv are no longer needed.

machacekondra replied:
done

pkliczewski commented:
@goneri Any news? Did you have some time to retest?

pkliczewski commented:
@mattclay @bcoca Please review whether anything you wanted was fixed.

goneri commented:
shipit

s-hertel commented on lib/ansible/plugins/inventory/kubevirt.py:
```diff
@@ -0,0 +1,263 @@
+
+    def _sanitize_group_name(self, name):
+        """ Replace unsupported charactes in group name by underscore character """
+        name = name.replace('/', '_')
```
Rather than overwriting the base method from BaseInventoryPlugin, why don't you remove this method? Then your calls to _sanitize_group_name will use a standard regex to better address all possible invalid characters.

s-hertel reviewed (changes requested):
One suggestion, but looks good to merge to me after addressed.

s-hertel committed:
Use standard _sanitize_group_name method

pkliczewski commented:
:tada: Thank you all for you help!!!
